Bring architecture doc up to current thinking. (#12)
AMWA-TV/ms-05-01: docs/Core Mechanisms.md

NCA defines a Controller as a software function, not as a physical entity, and a Device as a real or virtual construct. The command travel paths listed above refer to communication of information between software entities, not necessarily physical pieces of equipment.

> The actual means of control message transport and the formats of the messages are determined by the protocol being used. One NCA protocol, a JSON-based RPC scheme, is defined in [NC-Protocol](NC-Protocol.md).

Reviewer: ```suggestion
The actual means of control message transport and the formats of the messages are determined by the protocol being used. A JSON-based RPC protocol is available in the `NC-Protocol` repository.
```

> ## Control sessions ##

Reviewer: ```suggestion
## Control sessions
```

NCA control is session-oriented, which implies the following:

1. Requests and responses occur in correlated (but not blocking) pairs.
1. Controlled Devices (and their inventory of objects) have persistent application relationships with Controllers.
1. Prompt discovery of Device failure or communication failure is a required feature of the protocol.
1. Devices shall report changing parameters to subscribing Controllers. Subscriptions are valid through the conclusion of the control session unless explicitly un-subscribed.
1. When a communication failure occurs, the control session behaviour shall be well-defined. Whether the session survives or gracefully dies is a property of the protocol in use.

> ## Events, Notifications, and Subscriptions ##

Reviewer: ```suggestion
## Events, Notifications, and Subscriptions
```

> NCA supports the use of multiple simultaneous Controllers. Those Controllers might be coordinated - when a system has redundant Controllers, for example - or they might be uncoordinated - for example, when a system includes  Controllers from multiple makers.

Reviewer: ```suggestion
NCA supports the use of multiple simultaneous Controllers. Those Controllers might be coordinated - when a system has redundant Controllers, for example - or they might be uncoordinated - for example, when a system includes Controllers from multiple manufacturers.
```

A common application of multiple controllers arises when a network-controlled device includes a non-blank front panel.  The front panel is a separate Controller. Thus, every case in which NCA controls Devices that have their own physical controls is a multi-controller application.

In all multiple multi-controller situations, timely tracking of parameter and state updates is required for error-free Device control. NCA's core update-tracking mechanism is the **event.** An event is anything that happens inside a Device that a Controller may need to know about.

> Historically, Controllers have often tracked device changes by polling. A goal of NCA is to render polling completely unnecessary in all cases. To this end, NCA events generate **notifications.**  A notification is a spontaneous message from a Device to one or more Controllers that announces the occurrence of an event and gives its particulars.

Reviewer: ```suggestion
Historically, Controllers have often tracked device changes by polling. A goal of NCA is to render polling completely unnecessary in all cases. To this end, NCA events generate **notifications**.  A notification is a spontaneous message from a Device to one or more Controllers that announces the occurrence of an event and gives its particulars.
```

Not all Controllers need to know about all events. When a Controller wants to be informed of the occurrence of a certain event in a certain object, it registers a **subscription** with the Device's _subscription manager,_ an instance of the class **ncaSubscriptionManager.**  A subscription causes a **notification message** to be delivered to the subscribing Controller whenever the designated event is raised in the designated object.

> Once registered, a subscription remains in effect for the life of the control session, unless the Controller cancels it or the Controller's connection fails.

Reviewer: ```suggestion
Once registered, a subscription remains in effect for the life of the control session, unless the Controller cancels it or the Controller's session ends.
```

> ### Notification messages ###

Reviewer: ```suggestion
### Notification messages
```

A notification message is essentially a callback command whose parameters depend on the type of event. However, unlike actual commands, a notification does not require a response.

The parameters of notifications depend on the type of event. In general, event parameters are designed to give full detail about the event, so that Controllers receiving events can act on them immediately, without having to query the Device further.

> ### The PropertyChanged event ###

Reviewer: ```suggestion
### The PropertyChanged event
```

The most common event is the **PropertyChanged** event.  **PropertyChanged**   is defined by the base class **ncaObject**, so it is inherited by every object. 

**PropertyChanged** is the key mechanism for NCA's support of multiple Controllers without polling. By subscribing to changes in the properties it cares about, a Controller can stay in sync with the device, even when those properties are being changed by other controllers. 

> ## Controller hierarchies ##

Reviewer: ```suggestion
## Controller hierarchies
```

> Some NCA-controlled media networks may contain thousands of devices. In such cases, it will not be practical for a single central Controller to maintain a separate control session with each device. To address this problem, indirect control using a hierarchy of Controllers may be implemented, with successive levels aggregating control functionality in ways appropriate to the applications.  NCA's core mechanisms are sufficient to support such configurations, but the exact implementation details are out of this specification's scope.

Reviewer: ```suggestion
Some NCA-controlled media networks may contain thousands of devices. In such cases, it will not be practical for a single central Controller to maintain a separate control session with each device. To address this problem, indirect control using a hierarchy of Controllers can be implemented, with successive levels aggregating control functionality in ways appropriate to the applications.  NCA's core mechanisms are sufficient to support such configurations, but the exact implementation details are out of this specification's scope.
```

> ## Concurrency control ##

Reviewer: ```suggestion
## Concurrency control
```

10. It is possible to lock an entire Device by locking its Device Manager object (see [Managers](Managers.md)), provided that none of the device's other objects are locked.
11. Locked objects may not be deleted.

The NCA lock mechanism is for transient programmatic use. It is not:

- An authorization/permission mechanism for controlling user access to Devices or their features;
- A reservation mechanism for securing user access to Devices or features at a later time; or
- An operations mechanism for restricting Device access during maintenance, or when problems are occurring, or for other facility management purposes.

> ## Capability enumeration ##

Reviewer: ```suggestion
## Capability enumeration
```

> To discover the detailed capabilities of a block, a Controller can enumerate its objects and signal paths.  As well, the Controller can discover details of the classes from which those objects are built.

Reviewer: ```suggestion
To discover the detailed capabilities of a block, a Controller can enumerate its objects and signal paths.  As well, the Controller can discover details of the classes from which those objects are built and the data types used.
```

> ### Object enumeration ###

Reviewer: ```suggestion
### Object enumeration
```

These functions return simple lists of signal-path descriptors. Each signal-path descriptor specifies the ports at each end of the signal path.

> ### Class discovery ###

Reviewer: ```suggestion
### Class discovery
```

If a Controller does not have _a priori_ knowledge of a device's class definitions, it may discover the details of those definitions using the NCA _class discovery_ mechanism.  This mechanism returns descriptors to the Controller for each class and datatype the device uses.  These descriptors are sufficient for the Controller to construct well-formed API calls to the device; however, they do not describe the semantics of such calls.

> ## Reliability ##

Reviewer: ```suggestion
## Reliability
```

NCA protocols (defined in future work) may support additional reliability features - for example, Ethernet-based protocols can support spanning-tree Ethernet and/or can use reliable data transport protocols such as TCP.

> ### Device supervision ###

Reviewer: ```suggestion
### Device supervision
```
